xnmt/search_strategy.py

```diff
@@ -2,125 +2,253 @@
 import numpy as np
 from collections import namedtuple
 
+import xnmt.batcher
 from xnmt.length_normalization import NoNormalization
-from xnmt.persistence import bare
+from xnmt.persistence import bare, Serializable, serializable_init, Ref, bare
 from xnmt.vocab import Vocab
 
+
 # Output of the search
-SearchOutput = namedtuple('SearchOutput', ['word_ids', 'attentions'])
-Hypothesis = namedtuple('Hypothesis', ['score', 'output', 'state'])
+# words_ids: list of generated word ids
+# attentions: list of corresponding attention vector of word_ids
+# score: a single value of log(p(E|F))
+# logsoftmaxes: a corresponding softmax vector of the score. score = logsoftmax[word_id]
+# state: a NON-BACKPROPAGATEABLE state that is used to produce the logsoftmax layer
+#        state is usually used to generate 'baseline' in reinforce loss
+SearchOutput = namedtuple('SearchOutput', ['word_ids', 'attentions', 'score', 'logsoftmaxes', 'state', 'mask'])
 
 class SearchStrategy(object):
   '''
-  A template class to generate translation from the output probability model.
+  A template class to generate translation from the output probability model. (Non-batched operation)
   '''
-  def generate_output(self, decoder, attender, output_embedder, dec_state, src_length=None, forced_trg_ids=None):
+  def generate_output(self, translator, dec_state,
+                      src_length=None, forced_trg_ids=None):
     """
     Args:
-      decoder (Decoder):
-      attender (Attender):
-      output_embedder (Embedder): target embedder
+      translator (Translator): a translator
       dec_state (MlpSoftmaxDecoderState): initial decoder state
       src_length (int): length of src sequence, required for some types of length normalization
       forced_trg_ids (List[int]): list of word ids, if given will force to generate this is the target sequence
     Returns:
-      Tuple[List[int],float]: (id list, score)
+      List[SearchOutput]: List of (word_ids, attentions, score, logsoftmaxes)
     """
     raise NotImplementedError('generate_output must be implemented in SearchStrategy subclasses')
 
-class GreedySearch(SearchStrategy):
+class GreedySearch(Serializable, SearchStrategy):
```

Comment:
Do you think we even need greedy search anymore? If it'd be useful that's fine, but we could also remove it in favor of just using beam search with beam size 1.

Reply:
I had implemented this originally for unit testing purposes. I guess it would be reasonable to move it into the respective unit testing module instead if it's not needed here?

Reply:
I could go either way. In that case let's not change it as part of this PR.